### DISCUSSION TITLE
New Glassbreak Sensors

### DISCUSSION THREAD

**Discussion by alvinchen1** [tsightler/ring-mqtt#322] [Feature Requests]
Howdy!  New official Ring Glassbreak sensors are shipping.  https://ring.com/products/glass-break-sensor. Looks like the device is available, but untested, in the unofficial Ring API ring-client-api (https://github.com/dgreif/ring/blob/1691caa5cc0275460ff2961ad67d08f5007b30a8/homebridge/README.md#supported-devices-via-ring-alarm-and-ring-smart-lighting-hubs under Glassbreak sensors).  

I've got a few of them and as expected when I start Ring-MQTT, when it is enumerating devices, the device is marked as unknown.

![image](https://user-images.githubusercontent.com/26409535/154717539-dbc98118-694c-47fb-bdab-d98780d791af.png)

Unsupported device: sensor.glassbreak

Anything I might be able to help supply or provide to help add these devices to support?  As mentioned in the ring-client-api there are three possible sensors that I think would make sense, similar to the supported Ring Motion Sensor

- Detect Glass Breaking (akin to Ring Motion Sensor Motion Detected? so binary_sensor, maybe device class "problem"?)
- Tamper (again, probably binary_sensor?)
- Battery (regular battery sensor?)
 

Anyway, happy to help and test if I can, you know maybe without breaking any of my windows, of course :)

Thanks for the awesome work!

**Comment by tsightler** [comment]
They are basically implemented as just another class of contact sensors.  Technically I could, and probably should, combine all binary sensors into a single class, but for legacy reasons, motion sensors are their own class even though they are implemented in the API exactly like contact sensors, with the only difference being the device class.

Anyway, the glassbreak sensors should theoretically work in the dev branch, but I don't own any, so they are untested to this point.  Feel free to give the dev branch a spin and let me know.

**Reply by alvinchen1** [reply]
Cool, I see the code change. I have never run a dev branch of an addon before. I've installed portainer and lightly tinkered around in there, but that's the extent of that. I am going to look how to do it and will supply feedback on it when I can.  Thanks!

**Reply by tsightler** [reply]
You can just use the BRANCH feature of the addon, in the addon configuration tab just change branch config option from "addon" to "dev" and restart the addon.  On startup it will pull down the latest dev code and dependencies.  If for some reason it doesn't work just change the config option back to "addon" and restart and it will revert to the code in the pre-built Docker image.

**Reply by alvinchen1** [reply]
Sweet!  Thanks, I was just about there and looking at exactly that configuration setting.  Also, read through a few other of the posts and stuff and here's the output of ring-device-data-cli device discovery, if that would be helpful for you.



```
{
          "adapterType": "zwave",
          "batteryLevel": 100,
          "batteryStatus": "full",
          "categoryId": 5,
          "commStatus": "ok",
          "commandTypes": {
            "communication-poll": {
              "requiresTrust": false
            },
            "detection-test-mode.start": {
              "requiresTrust": false
            },
            "reconfigure.start": {
              "requiresTrust": false
            },
            "update-node-neighbors.start": {
              "requiresTrust": false
            }
          },
          "deviceFoundTime": 1645197896614,
          "deviceType": "sensor.glassbreak",
          "firmwareUpdate": {
            "eligibility": "eligible",
            "state": "up-to-date"
          },
          "impulseTypes": {
            "clear": {
              "trusted": true
            },
            "comm.ping.received": {
              "trusted": false
            },
            "fault": {
              "trusted": true
            },
            "tampered": {
              "trusted": true
            },
            "tampered-cleared": {
              "trusted": true
            }
          },
          "lastCommTime": 1645203185208,
          "lastUpdate": 1645203185452,
          "linkQuality": "ok",
          "managerId": "zwave",
          "manufacturerName": "Ring",
          "name": "Dining Room Glass Break Sensor",
          "placement": "unassigned",
          "pollInterval": 0,
          "roomId": 8,
          "setupByPluginStatus": "complete",
          "setupByUserStatus": "complete",
          "subCategoryId": 0,
          "tags": [
            "kitted",
            "scanned"
          ],
          "tamperStatus": "ok",
          "zid": "b6173f3f-d419-uuid",
          "faulted": false,
          "testMode": "inactive"
        },
        {
          "adapterType": "zwave",
          "batteryLevel": 100,
          "batteryStatus": "full",
          "categoryId": 5,
          "commStatus": "ok",
          "commandTypes": {
            "communication-poll": {
              "requiresTrust": false
            },
            "detection-test-mode.start": {
              "requiresTrust": false
            },
            "reconfigure.start": {
              "requiresTrust": false
            },
            "update-node-neighbors.start": {
              "requiresTrust": false
            }
          },
          "deviceFoundTime": 1645202399622,
          "deviceType": "sensor.glassbreak",
          "firmwareUpdate": {
            "eligibility": "eligible",
            "state": "up-to-date"
          },
          "impulseTypes": {
            "clear": {
              "trusted": true
            },
            "comm.ping.received": {
              "trusted": false
            },
            "fault": {
              "trusted": true
            },
            "tampered": {
              "trusted": true
            },
            "tampered-cleared": {
              "trusted": true
            }
          },
          "lastCommTime": 1645203490367,
          "lastUpdate": 1645203490597,
          "linkQuality": "ok",
          "managerId": "zwave",
          "manufacturerName": "Ring",
          "name": "Living Room Glass Break Sensor",
          "placement": "unassigned",
          "pollInterval": 0,
          "roomId": 2,
          "setupByPluginStatus": "complete",
          "setupByUserStatus": "complete",
          "subCategoryId": 0,
          "tags": [
            "kitted",
            "scanned"
          ],
          "tamperStatus": "ok",
          "zid": "c465966a-2802-uuid",
          "faulted": false,
          "testMode": "inactive"
        },
        {
          "adapterType": "zwave",
          "batteryLevel": 100,
          "batteryStatus": "full",
          "categoryId": 5,
          "commStatus": "ok",
          "commandTypes": {
            "communication-poll": {
              "requiresTrust": false
            },
            "detection-test-mode.start": {
              "requiresTrust": false
            },
            "reconfigure.start": {
              "requiresTrust": false
            },
            "update-node-neighbors.start": {
              "requiresTrust": false
            }
          },
          "deviceFoundTime": 1645203530614,
          "deviceType": "sensor.glassbreak",
          "firmwareUpdate": {
            "eligibility": "eligible",
            "state": "up-to-date"
          },
          "impulseTypes": {
            "clear": {
              "trusted": true
            },
            "comm.ping.received": {
              "trusted": false
            },
            "fault": {
              "trusted": true
            },
            "tampered": {
              "trusted": true
            },
            "tampered-cleared": {
              "trusted": true
            }
          },
          "lastCommTime": 1645203915279,
          "lastUpdate": 1645203915529,
          "linkQuality": "ok",
          "managerId": "zwave",
          "manufacturerName": "Ring",
          "name": "Spider Bathroom Glass Break Sensor",
          "placement": "unassigned",
          "pollInterval": 0,
          "roomId": 11,
          "setupByPluginStatus": "complete",
          "setupByUserStatus": "complete",
          "subCategoryId": 0,
          "tags": [
            "kitted",
            "scanned"
          ],
          "tamperStatus": "ok",
          "zid": "2aeba5b1-e32d-uuid",
          "faulted": false,
          "testMode": "inactive"
        }
```

**Reply by alvinchen1** [reply]
Dev Channel worked and the devices now show up in discovery.

![image](https://user-images.githubusercontent.com/26409535/154736498-1afc0351-9336-45b2-aa45-cf20c4dd1b96.png)

But only getting the Battery, Info, and Tamper sensors (which all look great). No binary_sensor of class Problem.
 
![image](https://user-images.githubusercontent.com/26409535/154736851-2cf085d3-4e68-4054-8355-72d0fe4b028f.png)
 
So this maybe an issue at the ring-client-api level?

**Reply by tsightler** [reply]
I think it's just a stupid bug (missing a break in the case statement so it was falling through to the generic function).  Restart should pull in the fix.

**Reply by alvinchen1** [reply]
Worked!  Kick ass!

![image](https://user-images.githubusercontent.com/26409535/154739174-02780921-ff47-472c-9be4-775fb3fbbbe8.png)


I am going to **assume** that means it is working.  Not sure how I'm gonna test a glass shatter, but at least this seems to bubble up the devices and sensors.  Maybe I'll start drinking early today and smash a beer bottle or two to test... :)


Now, just gotta wait till we get issue #319 working for my new Floodlight Pro camera arriving tomorrow... Might hold off installing that till that gets resolved.  Thanks again!

**Reply by tsightler** [reply]
Honestly, I'm not sure what is causing that issue.  The returned API data from the Floodlight Pro looks nearly identical in most respects to all other Ring cameras and there's no model specific differences in ring-client-api.  The only thing that should be the case is that the device model would display as unknown because it's not in the list (below is the line from ring-client-api):

this.model = RingCameraModel[this.initialData.kind] || 'Unknown Model'

There's absolutely zero model specific code in the camera handler, it literally just looks at the data returned by the API and determines if given features are there or not, and ring-mqtt depends on those indicators.  There's been no other users reports that the Floodlight Pro doesn't work, in either project, and I can't image that there aren't some out there already, and that user as, so far, not provided the request debug logs from ring-mqtt, so I don't really have any way to look into it deeper as I don't plan on buying one any time soon as I'm not convinced I'm going to keep investing into an ecosystem that seems to be closing more in every product release, even if I do understand it from their business perspective.

So, I guess my point is, if you do install it, please report your success/failure and provide debug logs as that will likely move it along much more quickly as I'm not convinced this issue is just the single missing entry in ring-client-api or I think the camera wouldn't show up at all.

**Comment by SeraphimSerapis** [comment]
Ha, I was just about to open a topic and then saw this one. I'd also love to see support for this device! Please let me know if there's anything I could do to help.

**Reply by alvinchen1** [reply]
Looks like the devices are showing up but untested in Dev branch.  If you'd like to try it out...

> You can just use the BRANCH feature of the addon, in the addon configuration tab just change branch config option from "addon" to "dev" and restart the addon. On startup it will pull down the latest dev code and dependencies. If for some reason it doesn't work just change the config option back to "addon" and restart and it will revert to the code in the pre-built Docker image.

![image](https://user-images.githubusercontent.com/26409535/154740464-294f00b6-1fb5-410b-871b-34d59c3fb3b6.png)

**Reply by tsightler** [reply]
As noted above (see last reply) it already works in the dev branch, after that one minor bugfix anyway.

**Reply by alvinchen1** [reply]
@SeraphimSerapis, run Dev, arm your alarm, and smash a window.  Let us know... :)

**Reply by tsightler** [reply]
It should work, it's literally just a new binary_sensor device with a different class and they all use "faulted".  Other than detection and proper device class settings (I used "safety" because there's no glassbreak device class in Home Assistant) there's no actual new code.

**Reply by alvinchen1** [reply]
Yup @tsightler agree.  That totally makes sense, safety device type over problem device type.    When do we think dev might merge into master?  Based upon when you think that will deploy, I may or may not remain in dev.

**Reply by tsightler** [reply]
It will be a while as dev is currently primarily focused on upcoming major changes in config/state management, at least a few weeks I'd guess, maybe more since I have limited time available in the coming weeks.  Admittedly most of the basis for that work is already in there, so it might be quicker if I magically find an extra few hours to finish the cleanup of the main initialization routine, do all the required testing for the various install methods, and update the docs.

If it goes on too long I'll probably try to pull the device updates into the 4.9.x branch, especially if we find out there are changes needed to support the Floodlight Pro.  That being said, there's not much risk to running the dev branch, other than it could be broken at any given minute, but usually if this happens you can restart a few minutes later and it will be fixed because I try to keep it in working state since I run it 100% of the time on my dev instance and quite regularly on my production instance, especially if I want to do some long term testing.

**Reply by alvinchen1** [reply]
Cool.  Even though I don't need those break glass sensors for really anything right now, I'll stay on dev for a bit then.  Especially, if it would be helpful to get logs for any other work you might be doing.  I'm fine testing and living on the bleeding edge a while.  Also, any way you'd like me to mark this thread as 'resolved in dev', so we don't have other folks saying "Me too..." etc?  Just minimize the noise so you can focus on what you need to focus on?

**Reply by tsightler** [reply]
I've added a "queued for release" tag to this thread.

Also, if you do get that Pro installed, please let me know how it goes and provide some logs of needed, I'd like to get that resolved.

**Reply by SeraphimSerapis** [reply]
I gave this a try and things seem to work well on dev. Thanks for the great work, @tsightler
